File: ProgLang/Motzkin.py
#Assignment 1

#Due January 30 at noon, submit via blackboard.

#The folowing Python program generates Motzkin trees (made of leaves, nodes with one branch and nodes with two branches)

# Motzkin tree of size n
from pprint import pprint
def mot (n) :
  if n==0 : 
    yield 'leaf'
  else :
    for m in mot(n-1) :
      yield ('unary',m)
      
    for k in range(0,n-1) :    
      for l in mot(k) :
        for r in mot(n-2-k) :        
          yield ('binary',l,r)


def mot2(n):
    result = []
    if n == 0:
        result.append('leaf')
        return result
    for m in mot2(n-1):
        result.append('unary('+ m +')')
    
    for k in range(0,n-1):    
      for l in mot2(k) :
        for r in mot2(n-2-k) :        
          result.append('binary(' + l + ',' + r +')')
    return result
    
def test() :
  for n in range(6) :
    pprint(list(mot2(n))) 
    
print(test())
'''if __name__ == '__main__':
    test();'''
#Write a program that generates the same output in C/C++ or Java.

File: ProgLang/Motzkin.java
package ProgLang;

import java.util.ArrayList;
import java.util.Iterator;


public class Motzkin {

	String unaryReturn = new String();
	String binaryReturn = new String();
	ArrayList m = new ArrayList();
	ArrayList l = new ArrayList();
	ArrayList r = new ArrayList();
	public String mot(int n) {
		if( n == 0 ) {
			return "leaf";
		} else {
			m.add(mot(n-1));
			Iterator it = m.iterator();
			while(it.hasNext()) {
			   unaryReturn = "(unary," + it.next().toString() + ")";	
			}
			for( int j=0;j<n-1;j++) {
				l.add(mot(j));
				Iterator it_l = l.iterator();

				r.add(mot(n-2-j));
				Iterator it_r = r.iterator();

				while(it_l.hasNext()) {
					while( it_r.hasNext()) {
						binaryReturn = "(binary," + it_l.next() + "," +  it_r.next() + ")";
					}
				}
			}
			

		}
		yield  unaryReturn + binaryReturn;
		
	}
	/**
	 * @param args
	 */
	public static void main(String args[]) {
//		Iterator it = new Iterator();
		String s = new String();
		for ( int i = 2 ; i <= 3; i++) {
			Motzkin motzkin = new Motzkin();
			s = motzkin.mot(i);
			System.out.println(s);
		}

		
	}
}